src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,330 @@
+    }
+    
+    /**
+     * Send token request with grant_type as refresh_token to token endpoint for getting new access token. 
+     */
+    AuthenticationResult acquireTokenWithRefreshToken(final String refreshToken) 
+            throws AuthenticationException {
+        Logger.v(TAG, "Try to get new access token with the found refresh token.", 
+                mAuthRequest.getLogInfo(), null);
+        
+        // Check if network is available, if not throw exception. 
+        throwIfNetworkNotAvaliable();
+        
+        final AuthenticationResult result;
+        try {
+            final JWSBuilder jwsBuilder = new JWSBuilder();
+            final Oauth2 oauthRequest = new Oauth2(mAuthRequest, mWebRequestHandler, jwsBuilder);
+            result = oauthRequest.refreshToken(refreshToken);
+            if (result != null && StringExtensions.IsNullOrBlank(result.getRefreshToken())) {
+                Logger.i(TAG, "Refresh token is not returned or empty", "");
+                result.setRefreshToken(refreshToken);
+            }
+        } catch (final IOException | AuthenticationException exc) {
+            // Server side error or similar
+            Logger.e(TAG, "Error in refresh token for request:" + mAuthRequest.getLogInfo(),
+                    ExceptionExtensions.getExceptionMessage(exc), ADALError.AUTH_FAILED_NO_TOKEN,
+                    exc);
+
+            throw new AuthenticationException(
+                    ADALError.AUTH_FAILED_NO_TOKEN, ExceptionExtensions.getExceptionMessage(exc),
+                    exc);
+        }
+
+        return result;
+    }
+    
+    /**
+     * For testing purpose, inject the dependency. 
+     */
+    void setWebRequestHandler(final IWebRequestHandler webRequestHandler) {
+        this.mWebRequestHandler = webRequestHandler;
+    }
+    
+    /**
+     * Attempt to get new access token with regular RT. 
+     */
+    private AuthenticationResult tryRT() throws AuthenticationException {
+        final TokenCacheItem regularRTItem = mTokenCacheAccessor.getRegularRefreshTokenCacheItem(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), getUserFromRequest());
+        
+        if (regularRTItem == null) {
+            Logger.v(TAG, "Regular token cache entry does not exist, try with MRRT.");
+            return tryMRRT(); 
+        }
+        
+        // When MRRT is returned, we store separate entries for both regular RT entry and MRRT entry, 
+        // we should look for MRRT entry if the token in regular RT entry is marked as MRRT.
+        // However, the current cache implementation never mark the token stored in regular RT entry
+        // as MRRT. To support the backward compatibility and improve cache lookup, when successfully
+        // retrieved regular RT entry token and if the mrrt flag is false, check the existence of MRRT.
+        if (regularRTItem.getIsMultiResourceRefreshToken() || isMRRTEntryExisted()) {
```

Comment:
Done

---
In reply to: [63076541](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63076541) [](ancestors = 63076541)